File: Data Structures & Algorithms/binary-search/submission-1.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        auto left = nums.begin();
        std::size_t dist = nums.size();

        while (dist > 0) {
            std::size_t half = dist >> 1;
            auto mid = left + half;

            if (*mid < target) {
                dist -= half + 1;
                left = mid + 1;
            } else {
                dist = half;
            }
        }

        if (left == nums.end())
            return -1;
        
        return *left == target? left - nums.begin() : -1;
    }
};

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-1.py
class Solution:
    def search(self, nums: List[int], t: int) -> int:
        s, e = 0, len(nums) - 1

        while s <= e:
            diff = (e - s) // 2
            mid = s + diff
            if nums[mid] == t:
                return mid

            if nums[mid] < nums[e]:
                if nums[mid] < t <= nums[e]:
                    s = mid + 1
                else:
                    e = mid - 1
            else:
                if nums[s] <= t < nums[mid]:
                    e = mid - 1
                else:
                    s = mid + 1
        
        return -1